### PR TITLE
Fix DeepInsert issue with matching RelatedDescriptors to NestedItems.…

### LINKED ISSUE #2781: Client - DeepInsert fails if more than one navigation property is sent
*Handle response from DeepInsert fails if more than one navigation property is returned on the same object.*

### Assemblies affected

*OData Client 7.18.0*

### Reproduce steps

*Try sending both a Manufacturer and an Owner on a Car during DeepInsert().*

### Expected result

*Request succeeds without any error and the Car is returned together with an Owner an a Manufacturer*

### Actual result

*Fails during handling of the response as the logic for matching NestedItems with related EntityDescriptor is wrong. An exception is thrown.*

```
  Message: 
    System.ArgumentException : An item with the same key has already been added.

  Stack Trace: 
    ThrowHelper.ThrowArgumentException(ExceptionResource resource)
    Dictionary`2.Insert(TKey key, TValue value, Boolean add)
    DeepInsertSaveResult.CreateOperationResponse(MaterializerEntry materializerState, Boolean isTopLevelResponse, Descriptor descriptor, OperationResponse parentOperationResponse) line 486
    DeepInsertSaveResult.HandleDeepInsertResponseInternal(MaterializerEntry entry, Boolean isTopLevelDescriptor, Descriptor descriptor, OperationResponse parentOperationResponse) line 420
    DeepInsertSaveResult.HandleDeepInsertResponseInternal(MaterializerEntry entry, Boolean isTopLevelDescriptor, Descriptor descriptor, OperationResponse parentOperationResponse) line 440
    DeepInsertSaveResult.HandleDeepInsertResponseInternal(MaterializerEntry entry, Boolean isTopLevelDescriptor, Descriptor descriptor, OperationResponse parentOperationResponse) line 440
    DeepInsertSaveResult.HandleDeepInsertResponse() line 413
    DeepInsertSaveResult.HandleResponse() line 366
    BaseSaveResult.EndRequest() line 304
    DataServiceContext.DeepInsert[T](T resource) line 2359

```

### Additional detail

*Optional, details of the root cause if known. Delete this section if you have no additional details to add.*

### PR DESCRIPTION
**Pull request by uffelauesen**
Fixes issue #2781

### Issues

*This pull request fixes #2781*

### Description

*2 new tests added showing the issue. The fix handles counting across different NestedItems on same nesting level.*

### Checklist (Uncheck if it is not completed)

- [ x] *Test cases added*
- [ x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Review by xuzhg — APPROVED** [review]

**Review by KenitoInc — DISMISSED** [review]

**Comment by KenitoInc** [comment]
@uffelauesen 
Look into the failing test
```
Microsoft.Data.WebClient.UnitTests.ClientRegressionTests.ShouldAcceptResponse204InAsync
Test method Microsoft.Data.WebClient.UnitTests.ClientRegressionTests.ShouldAcceptResponse204InAsync threw exception: 
Microsoft.OData.Client.DataServiceTransportException: System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.WebException: Unable to connect to the remote server ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:23570
   at System.Net.Sockets.Socket.InternalEndConnect(IAsyncResult asyncResult)
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Net.Http.HttpClientHandler.GetResponseCallback(IAsyncResult ar)
   --- End of inner exception stack trace --- ---> System.Net.WebException: System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.WebException: Unable to connect to the remote server ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:23570
   at System.Net.Sockets.Socket.InternalEndConnect(IAsyncResult asyncResult)
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at System.Net.ServicePoint.ConnectSocketInternal(Boolean connectFailure, Socket s4, Socket s6, Socket& socket, IPAddress& address, ConnectSocketState state, IAsyncResult asyncResult, Exception& exception)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Net.Http.HttpClientHandler.GetResponseCallback(IAsyncResult ar)
   --- End of inner exception stack trace ---
```

**Comment by uffelauesen** [comment]
@KenitoInc 
Thanks for the heads up.
However, I can't see how any of my changes could lead to that test failing. The failing test looks like some dependent environment that fails, a timeout or some infrastructure issue.
I can see the Release pipeline is still executing, it has been so for > 1 day, is that normal?
/uffe

**Comment by uffelauesen** [comment]
@KenitoInc 
The mentioned test seems to be failing in other builds as well, like here...
https://github.com/OData/odata.net/pull/2505/checks?check_run_id=8213298113
Again - I don't think it is related to any of the changes in this PR.
/uffe

File: src/Microsoft.OData.Client/DeepInsertSaveResult.cs [hunk 1]
```diff
@@ -420,6 +420,7 @@ private void HandleDeepInsertResponseInternal(MaterializerEntry entry, bool isTo
             OperationResponse response = this.CreateOperationResponse(entry, isTopLevelDescriptor, descriptor, parentOperationResponse);
 
             List<Descriptor> relatedDescriptors = this.BulkUpdateGraph.GetRelatedDescriptors(descriptor);
+            int currentRelatedIndex = 0;
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.Client/DeepInsertSaveResult.cs:423]: add a comment of what this variable does or is.

File: src/Microsoft.OData.Client/DeepInsertSaveResult.cs [hunk 2]
```diff
@@ -431,24 +432,23 @@ private void HandleDeepInsertResponseInternal(MaterializerEntry entry, bool isTo
                         {
                             for (int i = 0; i < feed.Items.Count; i++)
                             {
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.Client/DeepInsertSaveResult.cs:434]: This loop handles all nested resources for a particular resource at that level.. Like if we have Product and product has Order and Customer related navigation properties.. It means we'll handle the Order and Customer at this level. What if Order and Customer had navigation properties? Could you add a test for that too?

**Review by KenitoInc — DISMISSED** [review]

**Comment by KenitoInc** [comment]
@uffelauesen Look into review comments

File: src/Microsoft.OData.Client/DeepInsertSaveResult.cs (hunk 1 above)
Comment (uffelauesen) [reply src/Microsoft.OData.Client/DeepInsertSaveResult.cs:423]: Will do

File: src/Microsoft.OData.Client/DeepInsertSaveResult.cs (hunk 2 above)
Comment (uffelauesen) [reply src/Microsoft.OData.Client/DeepInsertSaveResult.cs:434]: `foreach (IMaterializerState nestedItem in entry?.NestedItems)`
This is already part of the test as Manufactures and owners are both on the same level - the Car.

`for (int i = 0; i < feed.Items.Count; i++)`
This is when there a re multiple entries of the same multi navigation property. I can add another car to the test to showcase multiple cars on the same person.

Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/DeepInsertSaveResult.cs:435]: ok. I think I meant if Manufacturers and Owners had some nesting...like the nesting can go as deep as possible. There is no limit.

**Review by KenitoInc — APPROVED** [review]

**Comment by KenitoInc** [comment]
@uffelauesen Thanks for this PR

**Pull request merged**